app/cdap/components/ConnectionList/index.tsx

```diff
@@ -0,0 +1,271 @@
+        showTabs: false,
+        selectedTab: '',
+        isSearching: false,
+      });
+      if (res.entities) {
+        tempData[index + 1][`data`] = res.entities;
+      } else {
+        tempData[index + 1][`data`] = res;
+      }
+      tempData[index + 1][`showTabs`] = true;
+      tempData[index + 1][`selectedTab`] = null;
+      tempData[index + 1][`isSearching`] = false;
+      return tempData.slice(0, index + 2);
+    });
+  };
+
+  const getCategorizedConnectionsforSelectedTab = async (selectedValue: string, index: number) => {
+    const categorizedConnections = await getCategorizedConnections();
+    const connections = categorizedConnections.get(selectedValue) || [];
+    setDataForTabsHelper(connections, index);
+    toggleLoader(false);
+  };
+
+  const fetchEntities = async (connectionId, url = pathFromUrl) => {
+    const pathDesired = url ? url : pathFromUrl;
+    const entitiesPromise = exploreConnection({
+      connectionid: connectionId,
+      path: pathDesired,
+    });
+    return entitiesPromise.then((values) => {
+      return values;
+    });
+  };
+
+  const selectedTabValueHandler = (entity: IRecords, index: number) => {
+    toggleLoader(true);
+    setDataForTabs((currentData) => {
+      let newData = [...currentData];
+      newData[index].selectedTab = entity.name;
+      newData = newData.map((each) => {
+        return {
+          data: each.data,
+          showTabs: each.showTabs,
+          selectedTab: each.selectedTab,
+          isSearching: false,
+        };
+      });
+      if (index === 0) {
+        getCategorizedConnectionsforSelectedTab(entity.name, index);
+      } else if (index === 1) {
+        fetchEntities(entity.name).then((res) => {
+          setDataForTabsHelper(res, index);
+          toggleLoader(false);
+        });
+      } else {
+        if (entity.canBrowse) {
+          fetchEntities(dataForTabs[1].selectedTab, entity.path).then((res) => {
+            setDataForTabsHelper(res, index);
+            toggleLoader(false);
+          });
+        }
+      }
+      return newData;
+    });
+  };
+
+  useEffect(() => {
+    getConnectionsTabData();
+  }, []);
+
+  useEffect(() => {
+    setDataForTabs((prev) => {
+      const temp = prev;
+      temp[0].selectedTab = connectorType;
+      return temp;
+    });
+    getCategorizedConnectionsforSelectedTab(connectorType, 0);
+  }, [connectorType]);
+
+  const headerForLevelZero = () => {
+    return (
+      <Box className={classes.styleForLevelZero}>
+        <Typography variant="body2">Data Connections</Typography>
+      </Box>
+    );
+  };
+
+  let headerContent;
+
+  return (
+    <Box data-testid="data-sets-parent" className={classes.connectionsListContainer}>
+      <SubHeader />
+      <SelectDatasetWrapper>
+        {dataForTabs.map((each, index) => {
+          if (each.data.filter((el) => el.connectionId).length) {
+            connectionId = each.data.filter((el) => el.connectionId)[0].connectionId;
```

Comment:
Does this run the same array traversal twice in a row? Can this `filter` call be moved before the `if` so it's only run once?

Reply:
Done